lib/spack/spack/spec.py

```diff
@@ -4475,6 +4477,43 @@ def cformat(self, *args, **kwargs):
         kwargs.setdefault("color", None)
         return self.format(*args, **kwargs)
 
+    def format_path(self, format_string: str):
+        """Given a `format_string` that is intended as a path, generate a string
+        like from `Spec.format`, but eliminate extra path separators introduced by
+        formatting of Spec properties.
+
+        Path separators explicitly added to the string are preserved, so for example
+        "{name}/{version}" would generate a directory based on the Spec's name, and
+        a subdirectory based on its version; this function guarantees though that
+        the resulting string would only have two directories (i.e. that if under
+        normal circumstances that `str(Spec.version)` would contain a path
+        separator, it would not in this case).
+
+        Note that POSIX or Windows-style paths can be supplied on either POSIX or
+        Windows, and both "/" and "\" are treated as path separators (as a side
+        effect, "invalid" paths which mix these will be converted to OS-appropriate
+        paths with consistent separators).
+        """
+        format_component_with_sep = r"\{[^}]*[/\\][^}]*}"
+        if re.search(format_component_with_sep, format_string):
+            raise SpecFormatPathError(
+                f"Format component requests a '/', but format_path must convert: {format_string}"
```

Comment:
This message is a bit mysterious. What about "invalid format string for paths, cannot contain {/...}" or something?